ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,228 @@
+      List<IndexValue> validIndexValues =
+          getValidIndexValues(indexSegment, deleteReferenceTimeInMs, expirationReferenceTimeInMs, deletedKeys);
+      for (IndexValue value : validIndexValues) {
+        updateMapHelper(validSizePerLogSegment, indexSegment.getLogSegmentName(), value.getSize());
+      }
+    }
+    return validSizePerLogSegment;
+  }
+
+  /**
+   * Get a {@link List} of valid {@link IndexValue} in a given {@link IndexSegment}.
+   * @param indexSegment the indexSegment to be read from
+   * @param deleteReferenceTimeInMs the reference time in ms until which deletes are relevant
+   * @param expirationReferenceTimeInMs the reference time in ms until which expiration are relevant
+   * @param deletedKeys a {@link Set} of deleted keys to help determine whether a put is deleted
+   * @return a {@link List} of valid {@link IndexValue}
+   * @throws StoreException
+   */
+  private List<IndexValue> getValidIndexValues(IndexSegment indexSegment, long deleteReferenceTimeInMs,
+      long expirationReferenceTimeInMs, Set<String> deletedKeys) throws StoreException {
+    diskIOScheduler.getSlice(BlobStoreStats.IO_SCHEDULER_JOB_TYPE, BlobStoreStats.IO_SCHEDULER_JOB_ID, 1);
+    List<IndexEntry> indexEntries = new ArrayList<>();
+    List<IndexValue> validIndexValues = new ArrayList<>();
+    try {
+      indexSegment.getIndexEntriesSince(null, new FindEntriesCondition(Integer.MAX_VALUE), indexEntries,
+          new AtomicLong(0));
+    } catch (IOException e) {
+      throw new StoreException("I/O error while getting entries from index segment", e, StoreErrorCodes.IOError);
+    }
+    for (IndexEntry indexEntry : indexEntries) {
+      IndexValue indexValue = indexEntry.getValue();
+      if (indexValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
+        // delete record
+        validIndexValues.add(indexValue);
+        long operationTimeInMs =
+            indexValue.getOperationTimeInMs() == Utils.Infinite_Time ? indexSegment.getLastModifiedTimeMs()
+                : indexValue.getOperationTimeInMs();
+        if (operationTimeInMs < deleteReferenceTimeInMs) {
+          // delete is relevant
+          deletedKeys.add(indexEntry.getKey().getID());
+        } else if (!isExpired(indexValue.getExpiresAtMs(), expirationReferenceTimeInMs)
+            && indexValue.getOriginalMessageOffset() != -1
+            && indexValue.getOriginalMessageOffset() >= indexSegment.getStartOffset().getOffset()) {
+          // delete is irrelevant but it's in the same index segment as the put and the put is still valid
+          BlobReadOptions originalPut =
+              index.getBlobReadInfo(indexEntry.getKey(), EnumSet.of(StoreGetOptions.Store_Include_Deleted));
+          Offset originalPutOffset = new Offset(originalPut.getLogSegmentName(), originalPut.getOffset());
+          validIndexValues.add(new IndexValue(originalPut.getSize(), originalPutOffset, originalPut.getExpiresAtMs(),
```

Comment:
As of now the operation time is not needed by the caller and it's also not available in BlobReadOptions.